### PR TITLE
Add other_versions subfield for RPM packages.

### PR DESCRIPTION
**Pull request by jjustice6**
Signed-off-by: Joshua Justice <jjustice6@bloomberg.net>

## Description
It's legal to have two rpm packages with the exact same name installed. not (for example) python and python2.7, but literally the exact same name. Unfortunately, Ohai uses the package name as the key of the hash. This means Ohai currently only keeps the information of the last package in the list. 

This now produces a list of all versions of the package if there is more than one version present.

## Related Issue
https://github.com/chef/ohai/issues/1368

Please note that we strongly desire that this be backported into chef-client 14.x after its merge.

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/ohai/plugins/packages.rb [hunk 1]
```diff
@@ -50,7 +50,16 @@
 
       pkgs.each do |pkg|
         name, epoch, version, release, installdate, arch = pkg.split
-        packages[name] = { "epoch" => epoch, "version" => version, "release" => release, "installdate" => installdate, "arch" => arch }
+        if packages[name]
+          # We have more than one package with this exact name!
+          # Create an "other_versions" array for tracking all versions of packages with this name.
+          # The top-level package information will be the first one returned by rpm -qa,
+          # all others go in this list, with the same information they'd normally have.
+          packages[name]["other_versions"] = [] unless packages[name]["other_versions"]
```
Comment (tas50) [inline lib/ohai/plugins/packages.rb]:
```suggestion
          packages[name]["other_versions"] ||= []
```

File: lib/ohai/plugins/packages.rb [hunk 2]
```diff
@@ -50,7 +50,16 @@
...
+          packages[name]["other_versions"] << Mash.new({ "epoch" => epoch, "version" => version, "release" => release, "installdate" => installdate, "arch" => arch })
```
Comment (tas50) [inline lib/ohai/plugins/packages.rb]: Any chance any of these would ever be unique so we could just use them as a key for a hash?

File: lib/ohai/plugins/packages.rb (hunk 1 above)
Comment (jjustice6) [reply lib/ohai/plugins/packages.rb]: Ah, perfect.

File: lib/ohai/plugins/packages.rb (hunk 2 above)
Comment (jjustice6) [reply lib/ohai/plugins/packages.rb]: Not unless you want to concatenate stuff. By definition you need name+version+release+architecture to guarantee uniqueness. You could roll the dice with epoch or installdate, but those aren't _guaranteed_ unique (and I suspect there are edge cases where they break, like install discs or similar). An array seems like the least bad option.

File: lib/ohai/plugins/packages.rb [hunk 3]
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (tas50) [inline lib/ohai/plugins/packages.rb]: shouldn't this be encoding?

Comment (lamont-granquist) [reply lib/ohai/plugins/packages.rb]: unnecessary now (this is pre-2.0 i think)

Comment (tas50) [reply lib/ohai/plugins/packages.rb]: Don't you still want it if you're trying to force the encoding.

Comment (lamont-granquist) [reply lib/ohai/plugins/packages.rb]: only if you want to set it to something other than utf-8

**Comment by lamont-granquist** [comment]
i think this would be more useful as `versions` rather than `other_versions` and just list all of them, which can then be used to sidestep the 'best' one to use as the 'main' one.

File: lib/ohai/plugins/packages.rb (hunk 3 above)
Comment (jjustice6) [reply lib/ohai/plugins/packages.rb]: Did a `(setq ruby-insert-encoding-magic-comment nil)` and cleaned this up. Sorry.

**Comment by jjustice6** [comment]
I'm not picking a "best" one, but rather keeping the first one out of the list. If we move it into an array of versions, that extra level would break existing code that just assumes there's one package. (Also would make the interface for RPMs different than for everything else without doing work on those.)

**Comment by jjustice6** [comment]
Looks like the buildkite failures have nothing to do with this, incidentally.

File: lib/ohai/plugins/packages.rb (hunk 2 above)
Comment (lamont-granquist) [reply lib/ohai/plugins/packages.rb]: the NEVRA itself (name-epoch-version-release-arch) should be unique

**Comment by tas50** [comment]
@jjustice6 if you rebase on master that build failure will go away

**Comment by jjustice6** [comment]
All right, there we go. I've also incorporated the return of the clobbering per discussion with @lamont-granquist in slack.

**Comment by jjustice6** [comment]
This approach now puts _all_ versions of a package in a version table, if and only if there are multiple versions of the package present (so no duplication of information in the majority of cases).

**Comment by jjustice6** [comment]
Do you have any further requests for this PR?

**Comment by glensc** [comment]
the final version adds `versions` subkey. the title should be updated. also this got merged into 14.8.14 as well (#1386)

**Comment by jjustice6** [comment]
@glensc Thanks for the catch, I've updated the title of this for posterity's sake.

**Pull request merged**